pkg/services/ngalert/store/loki_range_to_instant.go

```diff
@@ -19,57 +19,74 @@ func (t dsType) isLoki() bool {
 	return t.DS.Type == datasources.DS_LOKI
 }
 
-func canBeInstant(r *models.AlertRule) bool {
+// canBeInstant checks if any of the query nodes that are loki range queries can be migrated to instant queries.
+// If any are migratable, those indices are returned.
+func canBeInstant(r *models.AlertRule) ([]int, bool) {
 	if len(r.Data) < 2 {
-		return false
-	}
-	// First query part should be range query.
-	if r.Data[0].QueryType != "range" {
-		return false
-	}
-
-	var t dsType
-	// We can ignore the error here, the query just won't be optimized.
-	_ = json.Unmarshal(r.Data[0].Model, &t)
-
-	if !t.isLoki() {
-		return false
-	}
-
-	exprRaw := make(map[string]interface{})
-	if err := json.Unmarshal(r.Data[1].Model, &exprRaw); err != nil {
-		return false
-	}
-
-	// Second query part should be and expression.
-	if !expr.IsDataSource(r.Data[1].DatasourceUID) {
-		return false
+		return nil, false
 	}
+	var (
+		optimizableIndices []int
+		canBeOptimized     = false
+	)
+	// Loop over query nodes to find all Loki range queries.
+	for i := range r.Data {
+		if r.Data[i].QueryType != "range" {
+			continue
+		}
+		var t dsType
+		// We can ignore the error here, the query just won't be optimized.
+		_ = json.Unmarshal(r.Data[i].Model, &t)
 
-	// Second query part should be "last()"
-	if val, ok := exprRaw["reducer"].(string); !ok || val != "last" {
-		return false
+		if !t.isLoki() {
+			continue
+		}
+		var reduceIndex int
+		// Loop over all query nodes to find the reduce node.
+		for ii := range r.Data {
```

Comment:
This nested loop might as well be extracted into a cache above, something like:

```GO
// Loop over query nodes to cache all last reduce expressions.
lastExpressionCache := make(map[string]bool)
for i := range r.Data {
if !expr.IsDataSource(r.Data[i].DatasourceUID) {
	continue
}
exprRaw := make(map[string]interface{})
if err := json.Unmarshal(r.Data[i].Model, &exprRaw); err != nil {
	continue
}
// Second be "last()"
if val, ok := exprRaw["reducer"].(string); !ok || val != "last" {
	continue
}

if ref, ok := exprRaw["expression"].(string); ok {
	lastExpressionCache[ref] = true
}
}
```

Followed by:

```GO
if _, ok := lastExpressionCache[r.Data[i].RefID]; ok {
	canBeOptimized = true
	optimizableIndices = append(optimizableIndices, i)
}
```

Reply:
This is where I think we should consider moving the optimization code to the `expr` package as we can do this optimization on the parsed execution graph rather than having to re-create the graph ourselves.

Reply:
This shouldn't be part of this change. We are talking about slices that have most of the time less than five items. We can change that once we move it to the expression package.